SPEC.md

```diff
@@ -201,25 +201,157 @@ Note that logging mechanisms other than stdout and stderr are not required by th
+An executor MAY implement a "strict mode" where an image cannot run unless all isolators are in place.
 
-* uint: base 10 formatted unsigned int as a string
-* bytes: Suffix to a base 10 int to make it a K, M, G, or T for base 1024
+### Linux Isolators
+
+These isolators are specific to the Linux Kernel and are impossible to represent as a 1-to-1 mapping on other Kernels.
+The first example is "capabilities" but this will be expanded to include things such as SELinux, SMACK or AppArmor.
+
+#### linux/capabilities-remove-set
+
+* Scope: app
+
+**Parameters:**
+* **set** list of capabilities that will be removed from the processes's capabilities bounding set
+
+```
+"name": "linux/capabilities-remove-set",
+"set": [
+	"CAP_SYS_PTRACE",
+]
+```
+
+#### linux/capabilities-retain-set
+
+* Scope: app
+
+**Parameters:**
+* **set** list of capabilities that will be retained in the processes's capabilities bounding set
+
+```
+"name": "linux/capabilities-retain-set",
+"set": [
+	"CAP_KILL",
+	"CAP_CHOWN"
+]
+```
+
+### Resource Isolators
+
+#### resource/cpu
+
+* Scope: app/container
+
+**Parameters:**
+* **request** milli-cores that are requested
+* **limit** milli-cores that can be consumed before the Kernel unschedules
+
+```
+"name": "resource/cpu",
+"request": "250",
+"limit": "500"
+```
+
+**Note**: a milli-core is the milli-seconds/second that the app will be able to run. e.g. 1000 would represent full use of a single CPU core.
+
+#### resource/memory
+
+* Scope: app/container
+
+**Parameters:**
+* **request** bytes of memory that the app is requesting to use and will be pushed back to as a result of contention
+* **limit** bytes of memory that the app can allocate before the Kernel stops allowing allocations
+
+```
+"name": "resource/memory",
+"request": "1G",
+"limit": "2G"
+```
+
+#### resource/block-weight
+
+* Scope: app/container
+
+**Parameters:**
+* **devicePaths** the block devices that this weight will be placed on, if this is omitted then this weight will be the default for all devices
+* **request** read/write fraction of device I/O time between 1 and 100 allocated to this container
+
+```
+"name": "resource/block-weight",
+"default: false,
+"devicePaths": [
+	"/dev/sda"
+],
+"request": {
+	"read": "100",
+	"write": "100"
+}
+```
+
+#### resource/block-bandwidth
+
+* Scope: app/container
+
+**Parameters:**
+* **devicePaths** the block devices that this limit will be placed on
+* **limit** read/write bytes per second
+
+```
+"name": "resouce/block-bandwidth",
+"devicePaths": [
+	"/dev/sda"
+],
+"limit": {
+	"read": "2MB",
+	"write": "512MB"
+}
+```
+
+#### resource/block-iops
+
+* Scope: app/container
+
+**Parameters:**
+* **devicePaths** the block devices that this limit will be placed on
+* **limit** read/write iops per second
+
+```
+"name": "resource/block-iops",
+"devicePaths": [
+	"/dev/sda",
+],
+"limit": {
+	"read": "2000",
+	"write": "512"
+}
+```
+
+#### resource/network-bandwidth
+
+* Scope: app/container
+
+**Parameters:**
```

Comment:
Maybe specify device? That would assume a container can have multiple devices. May be simpler to assume only one for now.